teuthology/openstack/__init__.py

```diff
@@ -0,0 +1,621 @@
+                    continue
+                except Exception as e:
+                    if 'Unknown server' not in str(e):
+                        log.exception('cloud_init_wait ' + name_or_ip)
+                    client.close()
+                    if 'Unknown server' in str(e):
+                        continue
+                    else:
+                        raise e
+                log.debug('cloud_init_wait ' + all_done)
+                try:
+                    stdin, stdout, stderr = client.exec_command(all_done)
+                    stdout.channel.settimeout(5)
+                    out = stdout.read()
+                    log.debug('cloud_init_wait stdout ' + all_done + ' ' + out)
+                except socket.timeout as e:
+                    client.close()
+                    log.debug('cloud_init_wait socket.timeout ' + all_done)
+                    continue
+                except socket.error as e:
+                    client.close()
+                    log.debug('cloud_init_wait socket.error ' + str(e) + ' ' + all_done)
+                    continue
+                log.debug('cloud_init_wait stderr ' + all_done +
+                          ' ' + stderr.read())
+                if stdout.channel.recv_exit_status() == 0:
+                    success = True
+                client.close()
+                if success:
+                    break
+            return success
+
+    def exists(self, name_or_id):
+        """
+        Return true if the OpenStack name_or_id instance exists,
+        false otherwise.
+        """
+        servers = json.loads(misc.sh("openstack server list -f json"))
+        for server in servers:
+            if (server['ID'] == name_or_id or server['Name'] == name_or_id):
+                return True
+        return False
+
+    @staticmethod
+    def get_addresses(instance_id):
+        """
+        Return the list of IPs associated with instance_id in OpenStack.
+        """
+        with safe_while(sleep=2, tries=30,
+                        action="get ip " + instance_id) as proceed:
+            while proceed():
+                instance = misc.sh("openstack server show -f json " +
+                                   instance_id)
+                addresses = OpenStack.get_value(json.loads(instance),
+                                                'addresses')
+                found = re.match('.*\d+', addresses)
+                if found:
+                    return addresses
+
+    def get_ip(self, instance_id, network):
+        """
+        Return the private IP of the OpenStack instance_id. The network,
+        if not the empty string, disambiguate multiple networks attached
+        to the instance.
+        """
+        return re.findall(network + '=([\d.]+)',
+                          self.get_addresses(instance_id))[0]
+
+class TeuthologyOpenStack(OpenStack):
+
+    def __init__(self, args, config, argv):
```

Comment:
What is the difference between ``args`` and ``argv`` here? A docstring might be useful for this method.